core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1142,6 +1151,9 @@ private RexNode simplifyIs(RexCall call, RexUnknownAs unknownAs) {
     // "(CASE WHEN FALSE THEN 1 ELSE 2) IS NOT NULL" we first simplify the
     // argument to "2", and only then we can simplify "2 IS NOT NULL" to "TRUE".
     a = simplify(a, UNKNOWN);
+    if (!isSafeExpression(a)) {
+      return rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL, a);
+    }
     if (!a.getType().isNullable() && isSafeExpression(a)) {
```

Comment:
If we check the safety above we can remove the check from here since it will always true at this stage.